pom.xml

```diff
@@ -55,7 +55,7 @@
         <dependency>
             <groupId>org.apache.sling</groupId>
             <artifactId>org.apache.sling.api</artifactId>
-            <version>2.4.0</version>
+            <version>2.22.0</version>

```

Comment:
I would not change this one. While the parent bring is general improvements and the sling-mock is probably useful and brings in new features, updating the Sling API will also upgrade it for projects that consume this testing dependency.

We encourge projects to use the lowest compatible version of their dependencies so they can be deployed on as many platforms as possible.

Reply:
I believe I had to update this because of the upgrade to sling-mock 3.4.14.  That version of the project depends on sling.api 2.22.0

Without upgrading the sling.api 2.22.0, the project will not compile because of a java.lang.NoClassDefFoundError: org/apache/sling/api/wrappers/DeepReadModifiableValueMapDecorator error.

While I understand the sentiment behind using the lowest compatible version, I'm not sure I agree with it.  In practical terms, this doesn't really affect existing users.  If someone can't or doesn't want to upgrade their sling.api dependency, then they can always remain on the older version of this library.  If they want or need to use the latest version, then the price of admission is updating their dependencies.  No-one is forcing this on anyone.

Not upgrading though (for whatever reason) should not be a best practice (at least IMHO, which I admit lacks experience in the sling project).  Everyone should be upgrading their dependencies regularly.  If upgrading is a problem, then delay the upgrade until you can figure out what else needs to be changed, but I see upgrading eventually is inevitable so there's no point is putting it off too long.

My preference would be to not enable the bad practice of having outdated dependencies.

Having said that, if you feel strongly about this, it will take some work but I can keep rolling things back until I have a version that works, but I won't feel good about it. 😄

Reply:
Thanks for the detailed explanation @rmcdouga . In this particular case it makes perfect sense, since the Sling Mock pulls in version 2.22.0 of the Sling API.

In general, we do keep the dependency version as low as possible because this allows consumers to upgrade easily. We do want them to be able to upgrade as easy as possible, and having large batch upgrade of multiple versions is IMO to be avoided.